sys/mips/82371AB.c

```diff
@@ -0,0 +1,148 @@
+/* 82371AB PCI-ISA bridge driver */
+
+#define KL_LOG KL_DEV
+#include <sys/mimiker.h>
+#include <sys/device.h>
+#include <sys/rman.h>
+#include <sys/bus.h>
+#include <mips/malta.h>
+#include <sys/pci.h>
+#include <sys/interrupt.h>
+#include <dev/piixreg.h>
+#include <dev/isareg.h>
+#include <sys/libkern.h>
+#include <sys/devclass.h>
+
+typedef struct intel_isa_state {
+  resource_t *io;
+  rman_t io_rman;
+} intel_isa_state_t;
+
+static resource_t *intel_isa_alloc_resource(device_t *dev, res_type_t type,
+                                            int rid, rman_addr_t start,
+                                            rman_addr_t end, size_t size,
+                                            rman_flags_t flags) {
+  assert(dev->bus == DEV_BUS_ISA);
+
+  rman_t *rman = NULL;
+  intel_isa_state_t *isa = dev->parent->state;
+  bus_space_handle_t bh = 0;
+
+  if (type == RT_IOPORTS) {
+    assert(start >= IO_ICUSIZE);
+    rman = &isa->io_rman;
+    bh = isa->io->r_bus_handle;
+  } else {
+    /* We don't know how to allocate this resource, but maybe someone above us
+     * does. In current context this is necessary to allocate interrupts,
+     * which aren't managed by this driver. */
+    return bus_alloc_resource(dev->parent, type, rid, start, end, size, flags);
+  }
+
+  resource_t *r =
+    rman_reserve_resource(rman, type, rid, start, end, size, 0, flags);
+
+  if (r == NULL)
+    return NULL;
+
+  /* We shouldn't do that for IRQs, but rn we'll never end up here when
+   * allocating a resouce */
+  r->r_bus_tag = generic_bus_space;
+  r->r_bus_handle = bh + resource_start(r);
+
+  if (flags & RF_ACTIVE) {
+    if (bus_activate_resource(dev, r)) {
+      resource_release(r);
+      return NULL;
+    }
+  }
+
+  return r;
+}
+
+static int intel_isa_activate_resource(device_t *dev, resource_t *r) {
+  if (r->r_type != RT_IOPORTS)
+    return bus_activate_resource(dev->parent, r);
+  /* There are no resource activation procedures for ISA ioports  */
+  return 0;
+}
+
+static void intel_isa_deactivate_resource(device_t *dev, resource_t *r) {
+  if (r->r_type != RT_IOPORTS)
+    return bus_deactivate_resource(dev->parent, r);
+  return;
+}
+
+static void intel_isa_release_resource(device_t *dev, resource_t *r) {
+  if (r->r_type != RT_IOPORTS) {
+    /* Dispatch release for inderectly allocated resources */
+    bus_deactivate_resource(dev->parent, r);
+  } else {
+    /* No need to deactivate resource, because deactivation is a NOP anyway */
```

Comment:
You must call the deactivation function because it degrades the `range` activation counter. You're right that ` intel_isa_deactivate_resource` does not do anything, but `bus_deactivate_resource` does and it should be called.

Reply:
Done. Now it's always being called.